File: clients/python/src/aod/resources/environments.py
from __future__ import annotations

from typing import Any
from uuid import UUID

import httpx

from .._http import check_response
from ..models import Environment, EnvironmentVersion, Networking, NetworkingInput


def _normalize_networking(networking: NetworkingInput | None) -> dict[str, Any] | None:
    if networking is None:
        return None
    if isinstance(networking, Networking):
        return networking.model_dump(exclude_none=True)
    return networking


def _create_body(
    *,
    name: str,
    packages: dict[str, list[str]] | None = None,
    env_vars: dict[str, str] | None = None,
    setup_script: str | None = None,
    networking: NetworkingInput | None = None,
) -> dict[str, Any]:
    body: dict[str, Any] = {"name": name}
    if packages is not None:
        body["packages"] = packages
    if env_vars is not None:
        body["env_vars"] = env_vars
    if setup_script is not None:
        body["setup_script"] = setup_script
    normalized = _normalize_networking(networking)
    if normalized is not None:
        body["networking"] = normalized
    return body


def _update_body(
    *,
    version: int,
    name: str | None = None,
    packages: dict[str, list[str]] | None = None,
    env_vars: dict[str, str] | None = None,
    setup_script: str | None = None,
    networking: NetworkingInput | None = None,
) -> dict[str, Any]:
    body: dict[str, Any] = {"version": version}
    if name is not None:
        body["name"] = name
    if packages is not None:
        body["packages"] = packages
    if env_vars is not None:
        body["env_vars"] = env_vars
    if setup_script is not None:
        body["setup_script"] = setup_script
    normalized = _normalize_networking(networking)
    if normalized is not None:
        body["networking"] = normalized
    return body


class Environments:
    def __init__(self, client: httpx.Client) -> None:
        self._client = client

    def list(self) -> list[Environment]:
        body = check_response(self._client.get("/environments"))
        return [Environment.model_validate(e) for e in body["data"]]

    def create(
        self,
        *,
        name: str,
        packages: dict[str, list[str]] | None = None,
        env_vars: dict[str, str] | None = None,
        setup_script: str | None = None,
        networking: NetworkingInput | None = None,
    ) -> Environment:
        body = _create_body(
            name=name,
            packages=packages,
            env_vars=env_vars,
            setup_script=setup_script,
            networking=networking,
        )
        return Environment.model_validate(
            check_response(self._client.post("/environments", json=body))
        )

    def get(self, environment_id: str | UUID) -> Environment:
        return Environment.model_validate(
            check_response(self._client.get(f"/environments/{environment_id}"))
        )

    def update(
        self,
        environment_id: str | UUID,
        *,
        version: int,
        name: str | None = None,
        packages: dict[str, list[str]] | None = None,
        env_vars: dict[str, str] | None = None,
        setup_script: str | None = None,
        networking: NetworkingInput | None = None,
    ) -> Environment:
        body = _update_body(
            version=version,
            name=name,
            packages=packages,
            env_vars=env_vars,
            setup_script=setup_script,
            networking=networking,
        )
        return Environment.model_validate(
            check_response(self._client.put(f"/environments/{environment_id}", json=body))
        )

    def archive(self, environment_id: str | UUID) -> Environment:
        return Environment.model_validate(
            check_response(self._client.post(f"/environments/{environment_id}/archive"))
        )

    def delete(self, environment_id: str | UUID) -> None:
        check_response(self._client.delete(f"/environments/{environment_id}/delete"))

    def versions(self, environment_id: str | UUID) -> list[EnvironmentVersion]:
        body = check_response(self._client.get(f"/environments/{environment_id}/versions"))
        return [EnvironmentVersion.model_validate(v) for v in body["data"]]


class AsyncEnvironments:
    def __init__(self, client: httpx.AsyncClient) -> None:
        self._client = client

    async def list(self) -> list[Environment]:
        body = check_response(await self._client.get("/environments"))
        return [Environment.model_validate(e) for e in body["data"]]

    async def create(
        self,
        *,
        name: str,
        packages: dict[str, list[str]] | None = None,
        env_vars: dict[str, str] | None = None,
        setup_script: str | None = None,
        networking: NetworkingInput | None = None,
    ) -> Environment:
        body = _create_body(
            name=name,
            packages=packages,
            env_vars=env_vars,
            setup_script=setup_script,
            networking=networking,
        )
        return Environment.model_validate(
            check_response(await self._client.post("/environments", json=body))
        )

    async def get(self, environment_id: str | UUID) -> Environment:
        return Environment.model_validate(
            check_response(await self._client.get(f"/environments/{environment_id}"))
        )

    async def update(
        self,
        environment_id: str | UUID,
        *,
        version: int,
        name: str | None = None,
        packages: dict[str, list[str]] | None = None,
        env_vars: dict[str, str] | None = None,
        setup_script: str | None = None,
        networking: NetworkingInput | None = None,
    ) -> Environment:
        body = _update_body(
            version=version,
            name=name,
            packages=packages,
            env_vars=env_vars,
            setup_script=setup_script,
            networking=networking,
        )
        return Environment.model_validate(
            check_response(await self._client.put(f"/environments/{environment_id}", json=body))
        )

    async def archive(self, environment_id: str | UUID) -> Environment:
        return Environment.model_validate(
            check_response(await self._client.post(f"/environments/{environment_id}/archive"))
        )

    async def delete(self, environment_id: str | UUID) -> None:
        check_response(await self._client.delete(f"/environments/{environment_id}/delete"))

    async def versions(self, environment_id: str | UUID) -> list[EnvironmentVersion]:
        body = check_response(await self._client.get(f"/environments/{environment_id}/versions"))
        return [EnvironmentVersion.model_validate(v) for v in body["data"]]

File: clients/typescript/src/resources/environments.ts
import type { HttpClient } from "../http.js";
import type { Environment, EnvironmentVersion, Networking } from "../types.js";

export interface EnvironmentCreateParams {
  name: string;
  packages?: Record<string, string[]>;
  env_vars?: Record<string, string>;
  setup_script?: string;
  networking?: Networking;
}

export interface EnvironmentUpdateParams {
  version: number;
  name?: string;
  packages?: Record<string, string[]>;
  env_vars?: Record<string, string>;
  setup_script?: string;
  networking?: Networking;
}

interface ListResponse<T> {
  data: T[];
}

export class Environments {
  constructor(private readonly http: HttpClient) {}

  async list(opts: { signal?: AbortSignal } = {}): Promise<Environment[]> {
    const body = await this.http.request<ListResponse<Environment>>(
      "GET",
      "/environments",
      { signal: opts.signal },
    );
    return body.data;
  }

  async create(
    params: EnvironmentCreateParams,
    opts: { signal?: AbortSignal } = {},
  ): Promise<Environment> {
    return this.http.request<Environment>("POST", "/environments", {
      body: stripUndefined(params),
      signal: opts.signal,
    });
  }

  async get(
    environmentId: string,
    opts: { signal?: AbortSignal } = {},
  ): Promise<Environment> {
    return this.http.request<Environment>("GET", `/environments/${environmentId}`, {
      signal: opts.signal,
    });
  }

  async update(
    environmentId: string,
    params: EnvironmentUpdateParams,
    opts: { signal?: AbortSignal } = {},
  ): Promise<Environment> {
    return this.http.request<Environment>("PUT", `/environments/${environmentId}`, {
      body: stripUndefined(params),
      signal: opts.signal,
    });
  }

  async archive(
    environmentId: string,
    opts: { signal?: AbortSignal } = {},
  ): Promise<Environment> {
    return this.http.request<Environment>(
      "POST",
      `/environments/${environmentId}/archive`,
      { signal: opts.signal },
    );
  }

  async delete(
    environmentId: string,
    opts: { signal?: AbortSignal } = {},
  ): Promise<void> {
    await this.http.request<null>("DELETE", `/environments/${environmentId}/delete`, {
      signal: opts.signal,
    });
  }

  async versions(
    environmentId: string,
    opts: { signal?: AbortSignal } = {},
  ): Promise<EnvironmentVersion[]> {
    const body = await this.http.request<ListResponse<EnvironmentVersion>>(
      "GET",
      `/environments/${environmentId}/versions`,
      { signal: opts.signal },
    );
    return body.data;
  }
}

function stripUndefined<T extends object>(obj: T): Partial<T> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(obj)) {
    if (v !== undefined) out[k] = v;
  }
  return out as Partial<T>;
}
